File: src/lib/pet.ts
"use client";

/**
 * Host-side wrapper for the pet overlay window (src-tauri/src/pet.rs).
 * Typed replacement for the kit's vanilla pet-client.js — routes through
 * desktop.ts, so every call no-ops safely off-desktop / during SSR.
 */

import { invoke, listen } from "./desktop";

export type PetEvent =
  | "timerStart"
  | "timerResume"
  | "timerPause"
  | "timerBreak"
  | "breakEnd"
  | "timerFinish"
  | "timerAbandon"
  | "hover";

/** Coarse timer phase — drives the overlay card styling + finished panel. */
export type PetPhase = "idle" | "running" | "paused" | "finished";

/** Named standard-row states shared by the Flowmate overlay and Codex v2 atlas. */
export type PetAnimationState =
  | "idle"
  | "working"
  | "running"
  | "running_left"
  | "running_right"
  | "waving"
  | "jumping"
  | "failed"
  | "waiting"
  | "review"
  | "reading"
  | "sitting"
  | "drag_left"
  | "drag_right";

export type PetActionPayload = Record<
  string,
  string | number | boolean | null
>;

export interface PetAction {
  label: string;
  /** Omit to render a close-only chip such as the existing break "OK". */
  action?: string;
  /** Forwarded unchanged through Rust and merged into pet://control. */
  payload?: PetActionPayload;
}

export interface PetSignal {
  /** Direct standard-row animation state. Look directions remain renderer-owned. */
  state?: PetAnimationState;
  /** High-level event mapped to an animation by pet.config.json. */
  event?: PetEvent;
  /** Coarse timer phase — styles the card and toggles the finished panel. */
  phase?: PetPhase;
  /** Card title — e.g. the task name. */
  source?: string;
  /** Card timer line — e.g. the elapsed "HH:MM:SS" countup. */
  detail?: string;
  /** Quote/reminder text — renders the floating bubble + pointing pose (Row 4). */
  quote?: string;
  /** Kind of quote — styles the bubble + picks display duration/actions. */
  quoteKind?: "chat" | "break" | "reminder";
  /** Explicit speech-bubble actions. Break nudges supply defaults when omitted. */
  actions?: PetAction[];
  /** Correlates a pet chat reply to the request that opened the bubble. */
  chatRequestId?: string;
  /** Show the timer-complete extend chips (+5/+10/+25/Finish) on the card. */
  showExtend?: boolean;
  /** True while an external AI agent is working — glows the AI tab (no log UI). */
  agentActive?: boolean;
  /** Live one-line agent status for the AI tab (e.g. "Claude running · 4m"). */
  agentSummary?: string;
  /**
   * Offline / queued-writes state for the pet card.
   *
   * Pet mode hides the main window, and SyncStatusBadge lives in the main
   * header — so the surface where users spend their focus time was the one
   * surface with no sync indicator at all. "blocked" means writes exhausted
   * their retries and need the user.
   */
  syncState?: "ok" | "offline" | "syncing" | "blocked";
  /** Count behind syncState, e.g. 3 queued writes. */
  syncPending?: number;
  /** Auto-dismiss speech after this many ms; omit for quoteKind defaults. */
  speechMs?: number;
  /**
   * Short audio cue synthesised by the overlay. Omit for silence.
   * `cheer` also fires confetti — reserved for a genuine completion.
   */
  sound?: "cheer" | "alert";
  /** Force a desktop notification regardless of event. */
  notify?: { title: string; body: string };
}

/** Open (or reveal) the always-on-top pet overlay, optionally at physical px. */
export const petOpen = (x?: number, y?: number) => invoke("pet_open", { x, y });

/** Close the pet overlay window. */
export const petClose = () => invoke("pet_close");

/**
 * Whether the overlay is currently on screen.
 *
 * `petSignal()` emits into the pet webview whether or not it is visible, so a
 * proactive message sent while the pet is hidden is silently lost. Anything that
 * spends a limited budget (see `selectPetIntervention`) must check this first.
 * Resolves false off-desktop and during SSR.
 */
export const petIsOpen = async (): Promise<boolean> =>
  (await invoke<boolean>("pet_is_open")) === true;

/**
 * Animate the pet + (on finish events) fire a native notification.
 *
 * Resolves to `true` only when the overlay was on screen for this signal. Any
 * caller spending a limited budget must gate on that: the emit itself succeeds
 * against a hidden webview, so an unchecked send is a message thrown away.
 */
export const petSignal = async (signal: PetSignal): Promise<boolean> =>
  (await invoke<boolean>("pet_signal", { signal })) === true;

/** Move the pet window (physical px). */
export const petSetPosition = (x: number, y: number) =>
  invoke("pet_set_position", { x, y });

/** Toggle mouse pass-through (true = decorative only). */
export const petSetClickthrough = (enabled: boolean) =>
  invoke("pet_set_clickthrough", { enabled });

/** Start/stop the OS cursor-polling thread (emits pet://cursor to the overlay). */
export const petTracking = (enabled: boolean) =>
  invoke("pet_tracking", { enabled });

/** Payload from the pet's legacy poke/open surface. */
export interface PetPokePayload {
  action?: "openApp" | "poke";
  at?: number;
}

/** Listen for pet open/poke events. Returns an unlisten function. */
export const onPetPoke = (handler: (payload: PetPokePayload) => void) =>
  listen<PetPokePayload>("pet://poke", (payload) => handler(payload ?? {}));

/** Payload from the pet's control buttons (play/pause, extend chips, snooze). */
export interface PetControlPayload {
  action: string;
  minutes?: number;
  taskId?: string;
  estimateMinutes?: number;
  interventionKey?: string;
  requestId?: string;
  at?: number;
  /**
   * Chat panel: `requestPetReply` | `dismissPetReply`.
   * Also: toggle, extend, complete, openApp, idle*, agentRun*, etc.
   */
  [key: string]: unknown;
}

/** Listen for the pet's control buttons. Returns an unlisten fn. */
export const onPetControl = (handler: (payload: PetControlPayload) => void) =>
  listen<PetControlPayload>("pet://control", (p) =>
    handler(p ?? { action: "toggle" })
  );

File: tauri-pet-overlay/src-tauri/pet.rs
//! Agent Pet — overlay window module for Tauri v2.
//!
//! Drop this file next to `lib.rs` as `src-tauri/src/pet.rs`, then in `lib.rs`:
//!
//! ```ignore
//! mod pet;
//!
//! tauri::Builder::default()
//!     .plugin(tauri_plugin_notification::init())
//!     .invoke_handler(tauri::generate_handler![
//!         pet::pet_open,
//!         pet::pet_close,
//!         pet::pet_signal,
//!         pet::pet_set_position,
//!         pet::pet_set_clickthrough,
//!     ])
//!     // ...
//! ```
//!
//! The whole feature is driven by ONE command: `pet_signal`. Call it from your
//! timer logic whenever something happens — it animates the pet and fires a
//! native notification when a session finishes. See README.md.

use serde::{Deserialize, Serialize};
use tauri::{AppHandle, Emitter, Manager, PhysicalPosition, WebviewUrl, WebviewWindowBuilder};
use tauri_plugin_notification::NotificationExt;

/// Window label of the pet overlay. Used by the capability file too.
pub const PET_LABEL: &str = "pet";

/// Overlay window size (logical px). The sprite is centered inside it.
const PET_W: f64 = 220.0;
const PET_H: f64 = 240.0;

/// Optional native-notification payload.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct NotifyOpts {
    pub title: String,
    pub body: String,
}

/// A single instruction for the pet. Every field is optional so callers send
/// only what they have.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct PetSignal {
    /// Direct animation state: idle | running | waving | jumping | failed | waiting | review.
    #[serde(default)]
    pub state: Option<String>,
    /// High-level event mapped by `pet.config.json` -> `events`:
    /// timerStart | timerResume | timerPause | timerBreak | timerFinish | timerAbandon.
    #[serde(default)]
    pub event: Option<String>,
    /// Short label for the pet's speech bubble, e.g. the client name.
    #[serde(default)]
    pub source: Option<String>,
    /// Detail line, e.g. the task name or the countdown "24:59".
    #[serde(default)]
    pub detail: Option<String>,
    /// Force a desktop notification regardless of `event`.
    #[serde(default)]
    pub notify: Option<NotifyOpts>,
}

/// Open (or reveal) the always-on-top pet overlay window.
/// Pass `x`/`y` (physical px) to place it, e.g. next to the mini-widget.
#[tauri::command]
pub fn pet_open(app: AppHandle, x: Option<f64>, y: Option<f64>) -> Result<(), String> {
    if let Some(win) = app.get_webview_window(PET_LABEL) {
        win.show().map_err(|e| e.to_string())?;
        let _ = win.set_always_on_top(true);
        return Ok(());
    }

    let mut builder =
        WebviewWindowBuilder::new(&app, PET_LABEL, WebviewUrl::App("pet.html".into()))
            .title("Agent Pet")
            .inner_size(PET_W, PET_H)
            .decorations(false)
            .transparent(true)
            .always_on_top(true)
            .skip_taskbar(true)
            .resizable(false)
            .shadow(false)
            .focused(false)
            .visible(false);

    if let (Some(x), Some(y)) = (x, y) {
        builder = builder.position(x, y);
    }

    let win = builder.build().map_err(|e| e.to_string())?;
    win.show().map_err(|e| e.to_string())?;
    Ok(())
}

/// Close the pet overlay window.
#[tauri::command]
pub fn pet_close(app: AppHandle) -> Result<(), String> {
    if let Some(win) = app.get_webview_window(PET_LABEL) {
        win.close().map_err(|e| e.to_string())?;
    }
    Ok(())
}

/// The one call you need: animate the pet and (on finish events) notify.
///
/// Examples — from your timer code:
///   pet_signal({ event: "timerStart",  source: "Acme Co", detail: "Logo draft" })
///   pet_signal({ event: "timerPause" })
///   pet_signal({ event: "timerFinish", detail: "Logo draft — done" })
///   pet_signal({ state: "running",     detail: "24:59" })   // live countdown text
#[tauri::command]
pub fn pet_signal(app: AppHandle, signal: PetSignal) -> Result<(), String> {
    // 1. Drive the animation + bubble inside the pet window.
    app.emit_to(PET_LABEL, "pet://state", &signal)
        .map_err(|e| e.to_string())?;

    // 2. Native desktop notification: explicit, or implied by a finish event.
    let notify = signal.notify.clone().or_else(|| match signal.event.as_deref() {
        Some("timerFinish") => Some(NotifyOpts {
            title: "Pomodoro complete".into(),
            body: signal
                .detail
                .clone()
                .unwrap_or_else(|| "Time for a break.".into()),
        }),
        Some("timerAbandon") => Some(NotifyOpts {
            title: "Timer stopped".into(),
            body: "Session ended early.".into(),
        }),
        _ => None,
    });

    if let Some(n) = notify {
        app.notification()
            .builder()
            .title(&n.title)
            .body(&n.body)
            .show()
            .map_err(|e| e.to_string())?;

        // Re-assert the overlay on top so the user actually sees the reaction.
        if let Some(win) = app.get_webview_window(PET_LABEL) {
            let _ = win.show();
            let _ = win.set_always_on_top(true);
        }
    }

    Ok(())
}

/// Move the pet window (physical px). Handy to dock it beside the mini-widget.
#[tauri::command]
pub fn pet_set_position(app: AppHandle, x: f64, y: f64) -> Result<(), String> {
    if let Some(win) = app.get_webview_window(PET_LABEL) {
        win.set_position(PhysicalPosition::new(x, y))
            .map_err(|e| e.to_string())?;
    }
    Ok(())
}

/// Toggle click-through. `true` = mouse passes through the pet to windows
/// behind it (pet becomes purely decorative; click/hover/drag stop working).
#[tauri::command]
pub fn pet_set_clickthrough(app: AppHandle, enabled: bool) -> Result<(), String> {
    if let Some(win) = app.get_webview_window(PET_LABEL) {
        win.set_ignore_cursor_events(enabled)
            .map_err(|e| e.to_string())?;
    }
    Ok(())
}
